Added Support for RFC 8414 OAuth 2.0 Authorization Server Metadata

rhamedy opened the pull request:
Fixes: gh-6500

<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->

<!--
Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).
-->

rhamedy commented:
@jzheaux 
As per our conversations. Please see the first version of my changes. A few things 

- I have repurposed the usage of `ClientType` enum and I am open to alternative suggestions (if any)
- Refactored the existing code to come up with `reusable` utility methods 
- The existing tests are not updated and runs ✅ ensuring that existing functionality is not broken
- Added a matching unit test for `OAuth2` (matching with OpenID) 
- Added tests using `RequestDispatcher` to simulate the compatibility section of RFC 8414 

As you might have noticed, the `refactored/utility` methods in `JwtDecoders.java` and `ClientRegistrations.java` are very similar. Any suggestion on how we could avoid this duplication? Would be nice to port some of the common methods where it's accessible by both classes. 

Lastly, this is not the final change and the javadocs/comments require some more work. Once we agree on a `optimal` solution then I will proceed with addressing the comments and anything else that's necessary 👍

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -142,6 +214,74 @@ private static OIDCProviderMetadata parse(String body) {
+		String wellknownPath = WELL_KNOWN_PATH + type.getIssuerIdentifier();
+
+		URI uri = URI.create(issuer);
+		String issuerPath = uri.getPath();
```
I think a lot of this URL work can be cleaned up with `UriComponentsBuilder`.

rhamedy replied:
Thanks for suggesting `UriComponentBuilder` I have used that, let me know what you think of recent changes.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -100,16 +164,23 @@
+			return rest.getForObject(composeServerMetadataUrl(issuer, type, true), String.class);
+		} catch(HttpClientErrorException ex) {
+			 if(ex.getStatusCode() == HttpStatus.NOT_FOUND) {
+				 return rest.getForObject(composeServerMetadataUrl(issuer, type, false), String.class);
```
Thanks for spotting this! I overlooked this. I should have at least had a check for `clientType`. Anyway, with my recent changes I have fixed this as `open id` and `oauth2` have their own dedicated methods.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -69,19 +102,50 @@
+
+	public static ClientRegistration.Builder fromOAuth2IssuerLocation(String issuer) {
+		String configuration = getConfiguration(issuer, ClientType.OAUTH2);
+		OIDCProviderMetadata metadata = parse(configuration);
```
Make sense. I did have to remove the `getUserInfoEndpoint` though as its not available in `AuthorizationServerMetadata` 👍

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -142,6 +214,74 @@ private static OIDCProviderMetadata parse(String body) {
+			 * "https://example.com/.well-known/oauth-authorization-server/issuer1/" and in absence of path in supplied
+			 * issuer the string URl would look like "https://example.com/.well-known/oauth-authorization-server"
+			 */
+			completePath = (wellknownPath + issuerPath == null || issuerPath.equals("/") ? "" : issuerPath);
```
The answer is `never`, I have refactored this too 👍

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -142,6 +214,74 @@ private static OIDCProviderMetadata parse(String body) {
+	}
+
+	private static String getTypeName(ClientType type) {
+		return type.equals(ClientType.OIDC) ? "OpenID Configuration" : "OAuth Configuration";
```
I agree 👍 I have refactored this. It does not exist.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -142,6 +214,74 @@ private static OIDCProviderMetadata parse(String body) {
+			 * the value of path retrieved from supplied issuer. In it's absence the string URL would look as follow
+			 * "https://example.com/.well-known/openid-configuration"
+			 */
+			if(firstAttempt) {
```
I threw away the idea of having a `compose url` method and instead decided to have to separate `getOpenIdConfiguration` and `getOAuth2Configuration` and that way the logic is now contained in each method.

rhamedy commented:
Hey @jzheaux 

Thanks for the review. The branching logic did feel a little messy. I have made some changes (please see the recent commit)

- Removed the `validateIssuer` and `validateIssuer` methods 
- Removed the `ClientType` enum
- Created `getOpenIdConfiguration` and `getOAuth2Configuration` methods that also contain logic of how to construct the url
- I was thinking of refactoring the `parse` method to something as shown at the bottom. For now, created dedicated methods. 
- Refactored the `getScopes` method to make it reusable for both `OpenId` and `OAuth` by passing the `Scope` instead of `Metadata` 

```
private static <T> T parse(String body, Class<T> clazz) {
		Assert.isTrue((clazz.equals(OIDCProviderMetadata.class) || clazz.equals(AuthorizationServerMetadata.class)),
				"clazz allowed types are OIDCProviderMetadata and AuthorizationServerMetadata.");
		try {
			return clazz.equals(OIDCProviderMetadata.class)
					? clazz.cast(OIDCProviderMetadata.parse(body))
					: clazz.cast(AuthorizationServerMetadata.parse(body));
		}
		catch (ParseException e) {
			throw new RuntimeException(e);
		}
	}
```

Note: Once we are 👍 with the implementation details, then I will do the same for `JwtDecoders`

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -142,6 +233,15 @@ private static OIDCProviderMetadata parse(String body) {
 		}
 	}
 
+	private static AuthorizationServerMetadata parseAuthorizationServerMetadata(String body) {
```
Yeah, these checked exceptions make the code a little noisy. You might consider:

```java
private interface ThrowingFunction<S, T, E extends Throwable> {
    T apply(S src) throws E;
}
```

And then:

```java
private static <T> T parse(String body, ThrowingFunction<String, T, ParseException> parser) {
    try {
        return parser.apply(body);
    } catch (ParseException e) {
        throw new RuntimeException(e);
    }
}
```

It's probably sixes, though it could help clean up the two public methods:

```java
public static ClientRegistration.Builder withOidcIssuerLocation(String issuer) {
    String configuration = getOpenIdConfiguration(issuer);
    OIDCProviderMetadata metadata = parse(configuration, OIDCProviderMetadata::parse);
    return withProviderConfiguration(metadata)
            .userInfoUri(metadata.getUserInfoEndpointURI().toASCIIString());
}

public static ClientRegistration.Builder withOAuth2IssuerLocation(String issuer) {
    String configuration = getOAuth2Configuration(issuer);
    AuthorizationServerMetadata metadata = parse(configuration, AuthorizationServerMetadata::parse);
    return withProviderConfiguration(metadata);
}

private static ClientRegistration.Builder withProviderConfiguration(AuthorizationServerMetadata metadata) {
    // ... everything except for user info uri
}
```

rhamedy replied:
That definitely simplifies the public methods. 
We might have to pass the `issuer` as well to `withProviderConfiguration`. I also had to update the error message for `issuer` mismatch to as follow 

```
throw new IllegalStateException("The Issuer \"" + metadataIssuer + "\" provided in the configuration metadata did "
					+ "not match the requested issuer \"" + issuer + "\"");
```

jzheaux replied:
That's correct. `withOAuth2IssuerLocation` was just a typo on my part.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -100,16 +113,94 @@
+				 String url = UriComponentsBuilder.fromUri(uri).replacePath(uri.getPath() + wellKnownPath).toUriString();
+				 return rest.getForObject(url, String.class);
+			 } else {
+				 throw new IllegalArgumentException(invalidIssuerMessage);
```
Good idea 👍

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -120,11 +211,11 @@ private static ClientAuthenticationMethod getClientAuthenticationMethod(String i
 
-	private static List<String> getScopes(OIDCProviderMetadata metadata) {
-		Scope scope = metadata.getScopes();
+	private static List<String> getScopesAsStringList(Scope scope) {
```
I cannot believe how I missed this 😞 I checked the `AuthorizationServerMetadata` to see if it implements and interface but, I overlooked that `OIDCProviderMetadata` subclassing `AuthorizationServerMetadata`. Thanks for pointing out 👍

rhamedy commented:
@jzheaux pushed the recent changes to both `ClientRegistrations` and `JwtDecoders`. I have not touched the tests, please let me know if you have any ideas to make them more effective/useful. I also added some `javadoc` and once again, feel free to leave some comments.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -70,7 +78,7 @@
 	 * </p>
 	 * <pre>
-	 * ClientRegistration registration = ClientRegistrations.fromOidcIssuerLocation("https://example.com")
+	 * ClientRegistration registration = ClientRegistrations.withOidcIssuerLocation("https://example.com")
```
Ups! Gonna change them back to `fromOidcIssuerLocation` as well as below `fromOAuth2IssuerLocation` in my final changes :+1:

jzheaux commented on oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/JwtDecodersTests.java:
```diff
@@ -134,18 +180,69 @@ public void issuerWhenRequestedIssuerIsUnresponsiveThenThrowsIllegalArgumentExce
 
-	private void prepareOpenIdConfigurationResponse(String body) {
+	private void prepareConfigurationResponse(String body) {
+		//this.issuer = createIssuerFromServer();
```
Make sure to clean this up in your final changes

rhamedy replied:
Good catch! I took care of it 👍

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoders.java:
```diff
@@ -15,26 +15,33 @@
  * <a href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig">OpenID Provider Configuration</a>.
  *
  * @author Josh Cummings
+ * @author Rafiullah Hamedy
```
There is also `ReactiveJwtDecoders` to take care of. 

In that case, you might consider moving `getOpenIdConfiguration` and `getOAuth2Configuration` to a package-private class, since `JwtDecoders` and `ReactiveJwtDecoders` are in the same package, though abstracting this out is something that I think can wait for another PR.

rhamedy replied:
After looking at the `ReactiveJwtDecoders`, I agree it does make sense to refactor and reuse. I would be happy to work on a subsequent PR for it whenever we decide to create the issue for it 👍

jzheaux commented:
We should be careful about changing error messages, but these changes appear reasonable to me.

rhamedy commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/ClientRegistrationsTest.java:
```diff
@@ -118,11 +123,45 @@ public void cleanup() throws Exception {
+	 * @throws Exception
+	 */
+	@Test
+	public void issuerWithPathForOIDCCompatibilityAllInformaitonThenSuccess() throws Exception {
```
Thanks for pointing out. I updated the test names and replace OAuth occurrences with `Oauth2` in test names 👍

rhamedy commented:
Hi @jzheaux, appreciate the feedback. There is always something new to learn. 

I have pushed my recent changes. Please let me know if you there is more room for improvement. I have added some javadoc and once feel free to point out areas that are confusing and need further clarification 👍

jzheaux commented:
@rwinch @jgrandja Because we initially discussed `fromOidcIssuerLocation` together to establish its name, and we are introducing an additional method of a similar nature, I'd like to get your feedback on this contract before proceeding further.

The contract this PR proposes is:

* `fromOidcIssuerLocation` - no contract changes here, but it will now include the [compatibility measures recommended in RFC 8414](https://tools.ietf.org/html/rfc8414#section-5)
* `fromOAuth2IssuerLocation` - a new method that will query [the standard endpoint outlined in RFC 8414](https://tools.ietf.org/html/rfc8414#section-3.1) to obtain the same information

rhamedy commented:
@jzheaux based on our recent conversation on #6765, I have updated the `ClientRegistrations.java` and `ClientRegistrationsTest.java` for now. Please let me know if you wish any changes to be made before I go ahead and update the `JWTDecoder` its unit test and update the javadocs.

rhamedy commented:
Hi @jzheaux, I wanted to pick your brain with regards to unit testing. Now that we have decided to make the added contract `fromIssuerLocation` query all the endpoints (i.e. original oidc path, recommended oidc path by RFC 8414 and Oauth2, that's 3) then it means that for each existing test we need to add another 3 tests to test all the paths? Does that make sense? 

For example, the following test was there from before 

```
@Test
	public void issuerWhenResponseIsTypicalThenReturnedDecoderValidatesIssuer() {
		prepareConfigurationResponse();
		this.server.enqueue(new MockResponse().setBody(JWK_SET));

		JwtDecoder decoder = JwtDecoders.fromOidcIssuerLocation(this.issuer);

		assertThatCode(() -> decoder.decode(ISSUER_MISMATCH))
				.isInstanceOf(JwtValidationException.class)
				.hasMessageContaining("This iss claim is not equal to the configured issuer");
	}
```
So we keep the existing test `as is` since it's enforcing the existing `fromOidcIssuerLocation` contract. For the newly introduced contract `fromIssuerLocation` we need the following tests

1. Test the original oidc v1 (i.e. same as above test) ensuring that `fromIssuerLocation` can be used in place of `fromOidcIssuerLocation`
2. Test the recommended oidc by RFC 8414 i.e. `/.well-known/openid-configuration/issuer` 
3. Test the `oauth2` by RFC 8414 

The 3 test cases would look as follow (refactored assertions) and using separate `mock responses` to simulate each case 

```
        @Test
	public void issuerWhenOidcV1ResponseIsTypicalThenReturnedDecoderValidatesIssuer() {
		prepareConfigurationResponseForOidcV1("issuer1");
		JwtDecoder decoder = JwtDecoders.fromIssuerLocation(this.issuer);
		assertResponseIsTypicalThenReturnedDecoderValidatesIssuer(decoder);
	}

	@Test
	public void issuerWhenOidcResponseIsTypicalThenReturnedDecoderValidatesIssuer() {
		prepareConfigurationResponseForOidc("issuer1");
		JwtDecoder decoder = JwtDecoders.fromIssuerLocation(this.issuer);
		assertResponseIsTypicalThenReturnedDecoderValidatesIssuer(decoder);
	}

	@Test
	public void issuerWhenOauth2ResponseIsTypicalThenReturnedDecoderValidatesIssuer() {
		prepareConfigurationResponseForOauth2("issuer1");
		JwtDecoder decoder = JwtDecoders.fromIssuerLocation(this.issuer);
		assertResponseIsTypicalThenReturnedDecoderValidatesIssuer(decoder);
	}

	private void assertResponseIsTypicalThenReturnedDecoderValidatesIssuer(JwtDecoder decoder) {
		assertThatCode(() -> decoder.decode(ISSUER_MISMATCH))
		.isInstanceOf(JwtValidationException.class)
		.hasMessageContaining("This iss claim is not equal to the configured issuer");
	}
```
If you notice the first line of each of the above 3 test is different and that's because we would like the server to `respond` with `200` depending on which `case` we are testing otherwise `respond` with `404`. Here is a sample 

```
private void prepareConfigurationResponseForOauth2(String path) {
		this.issuer = this.server.url(path).toString();
		String body = String.format(DEFAULT_RESPONSE_TEMPLATE, this.issuer, this.issuer);

		final Dispatcher dispatcher = new Dispatcher() {
			@Override
			public MockResponse dispatch(RecordedRequest request) throws InterruptedException {
				switch(request.getPath()) {
					case "/.well-known/oauth-authorization-server/issuer1":
						return buildSuccessMockResponse(body);
					case "/issuer1/.well-known/jwks.json":
						return buildSuccessMockResponse(JWK_SET);
				}
				return new MockResponse().setResponseCode(404);
			}
		};
		this.server.setDispatcher(dispatcher);
	}
```
The switch case for `prepareConfigurationResponseForOidc` and `prepareConfigurationResponseForOidcV1` are different. 

We do the same for all of the existing tests for both `JwtDecodersTest` and `ClientRegistrationsTest`. Does this make sense? It kinda feels like a huge volume of test and I wanted to run it by you in case I have overlooked something and we do not have to do all that?

jzheaux commented:
@rhamedy, it's a good question - addressing all permutations can get pretty hard to keep up with for really not that much benefit.

I think the critical logic to verify is 

1. Does the endpoint resolution logic work?, e.g. does it correctly fall back, and
2. Does it correctly parse the response into a `ClientRegistrations.Builder`.

I think these can be tested independently for the most part - I wouldn't imagine there being 3x tests, it'd probably be one (or some) tests to verify the fallback logic and another set of tests confirming the parsing logic against the oauth2 endpoint.

rhamedy commented:
Hi @jzheaux 

Please see the recent changes. I have refactored the `JwtDecoers`, added `unit tests` as well as `javadoc`. I am sure there will be changes to be made, please feel free to let me know. 

As far as `ClientRegistrations` is concerned, since we changed the `scope` of `fromIssuerLocation` to query all 3 endpoint, before we invoke `withProviderConfiguration` we need to know whether this is `Oidc` or `Oauth2` and that is because we will have to include `.userInfoUri(metadata.getUserInfoEndpointURI().toASCIIString());` and that means we have to either parse `AuthorizationServerMetadata` or `OidcProviderMetadata` depending on which `endpoint` responded successfully. 
I have implemented a solution for now (my recent changes) but, feel free to purpose a different one comes in mind. 

I have also added `javadoc` please feel free to point out any that need to be updated. 
Thanks for these reviews/feedback btw.

rhamedy commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java:
```diff
@@ -95,21 +181,118 @@
+	 * @return Map<String, Object> - Configuration Metadata from the given issuer
+	 */
+	private static Map<ProviderType, String> getIssuerConfiguration(String issuer, String... paths) {
+		Assert.notEmpty(paths, "paths cannot be empty or null.");
```
Might have to change that message to `paths cannot be null or empty.`

rhamedy commented:
The use of `Map<ProviderType, String>` adds quite a bit of `branching` logic that might make the code not easy to read, would be open to suggestions on ways if we could `avoid` the `if-checks` and still be able to make `fromIssuerLocation` query all three endpoints.

jzheaux commented:
@rhamedy, this is now merged into `master` via f5b7706942350476bcb61f0b2b960e706543bab7. Thanks for all your efforts on this, I think it turned out well.

I squashed your commits (which is why you see a different commit hash) and then simplifed some of the branching logic that you called out - see the polish commit of 1739ef8d3c75871367e8bf9a0daf46547abf6d45 for details.

rhamedy commented:
Whoa! :open_mouth: That's a massive `polish`! Thanks for that. This PR definitely went through some ups and downs and change of requirements but, at the end of the day like you said turned out pretty good and thanks to your polish commit, things looks much clearer.
